File: algorithms/python/排序数组/排序数组.py
# Source : https://leetcode.cn/problems/sort-an-array/description/
# Author : Yanan Wang
# Date   : 2023-08-21

# 自从leetcode加了特殊测试用例之后，经典的快速排序（分区分两块）解法已经会超时了
# 原因是：当所有元素均相同时，O(nlgn)会退化成O(n^2)
# 所以，本题采用"分区分三块"的快速排序：
#      -----------------------
#      ^          ^          ^ 
#     <key      ==key       >key     
# 当所以元素均相同时，nums会变成一个分区(==key)，然后就退出递归。
# 分三区的做法采用《75. 颜色分类》题目中的三指针思想:
#   在遍历nums时，用left, i, right将其划分为四个区域:
#       [0, left)     < key
#       [left, i)    == key
#       [i, right]    待检测区域
#       (right, n-1]  > key 
#   当遍历完成时，nums会变成三个区域：
#       [0, left)     < key
#       [left, right] == key
#       (right, n-1]  > key 
# 选key的方法采用等概率随机法（算法导论中已证明，该方法的渐进时间复杂度最接近于O(nlgn)）

class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        def quickSort(nums, low, high):
            if low >= high:
                return
            key = nums[random.randint(low, high)]
            left, right = low, high
            i = low
            while i <= right:
                if nums[i] < key:
                    nums[left], nums[i] = nums[i], nums[left]
                    left += 1
                    i += 1
                elif nums[i] == key:
                    i += 1
                else:
                    nums[i], nums[right] = nums[right], nums[i]
                    right -= 1
            quickSort(nums, low, left - 1)
            quickSort(nums, right + 1, high)


        quickSort(nums, 0, len(nums) - 1)
        return nums

File: algorithms/cpp/排序数组/排序数组.cpp
// Source : https://leetcode.cn/problems/sort-an-array/description/
// Author : Yanan Wang
// Date   : 2023-08-23

/**
 * 方法一：快速排序
 * Leetcode 更新之后会超时，原因是引入了新的测试用例“元素全都相等”
 * 当元素全都相等时，快速排序的时间复杂度会变为O(N^2) 
*/
class Solution {
public:
    int Partition(vector<int> &nums, int left, int right) {
        int key = left;
        while (left < right) {
            while (left < right && nums[right] >= nums[key] && --right);
            while (left < right && nums[left] <= nums[key] && ++left);
            if (left < right) {
                swap(nums[left], nums[right]);
            }
        }
        swap(nums[key], nums[left]);
        return left;
    }
    void QuickSort(vector<int> &nums, int left, int right) {
        if (left >= right) {
            return;
        }
        int key = Partition(nums, left, right);
        QuickSort(nums, left, key - 1);
        QuickSort(nums, key + 1, right);
    }
    vector<int> sortArray(vector<int>& nums) {
        QuickSort(nums, 0, nums.size() - 1);
        return nums;            
    }
};

/**
 * 方法二：快速选择
 * 解决了当测试用例元素全都相等时，快速排序的时间复杂度为O(N^2)的问题
 * 此时快速选择的时间复杂度为O(N)
*/
class Solution {
public:
    void QuickSelect(vector<int> &nums, int begin, int end) {
        if (begin >= end) {
            return;
        }
        // 等概率随机法选key
        int key = nums[begin + rand() % (end - begin + 1)];
        int i = begin;
        int left = begin - 1;
        int right = end + 1;
        while (i < right) {
            if (nums[i] < key) {
                swap(nums[++left], nums[i++]);
            } else if (nums[i] == key) {
                ++i;
            } else {
                swap(nums[i], nums[--right]);
            }
        }
        QuickSelect(nums, begin, left);
        QuickSelect(nums, right, end);
    }
    vector<int> sortArray(vector<int>& nums) {
        srand(time(nullptr));
        QuickSelect(nums, 0, nums.size() - 1);
        return nums;            
    }
};

/**
 * 方法三：归并排序
*/
class Solution {
public:
    vector<int> tmp;
    void MergeSort(vector<int> &nums, int left, int right) {
        if (left >= right) {
            return;
        }
        int mid = left + (right - left) / 2;
        MergeSort(nums, left, mid);
        MergeSort(nums, mid + 1, right);
        int i = left;
        int cur1 = left;
        int cur2 = mid + 1;
        while (cur1 <= mid && cur2 <= right) {
            tmp[i++] = nums[cur1] < nums[cur2] ? nums[cur1++] : nums[cur2++]; 
        }
        while (cur1 <= mid) {
            tmp[i++] = nums[cur1++];
        }
        while (cur2 <= right) {
            tmp[i++] = nums[cur2++];
        }
        for (i = left; i <= right; ++i) {
            nums[i] = tmp[i];
        }
    }
    vector<int> sortArray(vector<int>& nums) {
        tmp.resize(nums.size());
        MergeSort(nums, 0, nums.size() - 1);
        return nums;            
    }
};